drivers: sensors: bmi270: Add CRT feature support

peter-axe opened the pull request:
### Issue: https://github.com/zephyrproject-rtos/zephyr/issues/62514

### Description
This PR integrates enhancements to the existing bmi270 driver by introducing support for the Component Retrimming (CRT) calibration feature. This is achieved through the incorporation of a new module for the CRT calibration process and ensuring its seamless integration with the existing functionalities of the bmi270 driver.

1. drivers: sensors: bmi270: Add macros and function prototypes for CRT:
- adds the necessary macros and function prototypes to the bmi270 driver header in order to add support for the CRT(Component Retrimming) calibration feature.

2. drivers: sensors: bmi270: add bmi270_soft_reset method public:
- Moves the soft reset logic, previously embedded in the init method to a to a new public method:bmi270_soft_reset that is required by the crt module.

3.  drivers: sensors: bmi270: Adds CRT feature module:
- Creates a specific kconfig for the feature.
- Adds a method to run the crt process.
- Adds a method to program the bmi270 nvm.
- Adds a method to enable/disable gain compensation.

4. drivers: sensors: bmi270:Add CRT support to sensor_attr_set:
Allows the bmi270 CRT feature methods to be called through Zephyr's sensor api. Using the sensor_attr_set function, with the following args:
-SENSOR_ATTR_CALIBRATION-run the crt.
-BMI270_SENSOR_ATTR_NVM_PROG-save the CRT results in nvm.
-BMI270_SENSOR_ATTR_GAIN_COMP-enable/disable gain compensation.

5. drivers: sensors: bmi270:Add CRT support to sensor_attr_get:
- allows using sensor_attr_get with SENSOR_ATTR_CALIBRATION to retrieve the compensated user-gain updated data of gyroscope. Also, it allows checking if CRT has ever been performed.

### Usage

1. in your prj.conf, add the feature config:
`CONFIG_BMI270_CRT=y`
2. In the dts instance of bmi270, add "bosch,bmi270-base" to the compatible list
This was introduced in a previous PR https://github.com/zephyrproject-rtos/zephyr/pull/55275 in order to download the bmi270 feature blob, that allows using bmi270 features, such as CRT.
e.g.:
```
 bmi270@69 {
                compatible = "bosch,bmi270", "bosch,bmi270-base";
```
3. Use the  sensor_attr_set from Zephyr-s sensor API to use the CRT feature.
For additional info, please consult the specific function doxygen documentation.
 e.g. test :
```
#include <zephyr/drivers/sensor/bmi270.h>

/* Run CRT */
sensor_attr_set(dev, SENSOR_CHAN_GYRO_XYZ,SENSOR_ATTR_CALIBRATION, NULL);
/* Program NVM */
sensor_attr_set(dev, SENSOR_CHAN_GYRO_XYZ,BMI270_SENSOR_ATTR_NVM_PROG, NULL);
/*  Enable gyro gain compensation */
SENSOR_ATTR_SET_GAIN_COMPENSATION(dev, BMI270_GYR_GAIN_EN);
```
4. (Optional) Use sensor_attr_get with SENSOR_ATTR_CALIBRATION to retrieve the compensated user-gain updated data of gyroscope. And confirm if CRT has ever been performed.
```
   struct sensor_value usr_gain = {0};
   int ret = sensor_attr_get(dev, SENSOR_CHAN_GYRO_XYZ, SENSOR_ATTR_CALIBRATION, &usr_gain);
   if (ret != 0) {
	printf("sensor_attr_get SENSOR_ATTR_CALIBRATION failed w/ err %d\n", ret);
    }
    uint8_t usr_gain_x = usr_gain.val1 & 0xFF; 
    uint8_t usr_gain_y = (usr_gain.val1 >> 8) & 0xFF;
    uint8_t usr_gain_z = (usr_gain.val1 >> 16) & 0xFF;
    printf("User gain get: x:%d, y:%d, z:%d/n", usr_gain_x, usr_gain_y, usr_gain_z);
```

### Tests
- Ensured previous functionalities and provided example remained intact.
- Performing CRT and programming the NVM:
```
*** Booting Zephyr OS build zephyr-v3.4.0-4217-g341590545ac8 ***
Device 0x93e4 name is bmi270@68
<dbg> bmi270: write_config_file: writing config file base
<inf> bmi270: Gyroscope user gain correction, X: 0 Y: 0 Z: 0
<dbg> bmi270: bmi270_set_aps: advance power save mode set to: 0
<dbg> bmi270: bmi270_set_aps: advance power save mode already in the intended state
<dbg> bmi270: bmi270_feature_reg_rmw: Read feature reg[0x32]@1 = 0x0000
<dbg> bmi270: bmi270_feature_reg_rmw: Wrote feature reg[0x32]@1 = 0x0000
<wrn> bmi270: Ensure that the device is at rest during CRT execution!
<dbg> bmi270: bmi270_set_aps: advance power save mode already in the intended state
<dbg> bmi270: bmi270_feature_reg_rmw: Read feature reg[0x32]@1 = 0x0000
<dbg> bmi270: bmi270_feature_reg_rmw: Wrote feature reg[0x32]@1 = 0x0100
<inf> bmi270: CRT running...
<dbg> bmi270: bmi270_set_aps: advance power save mode already in the intended state
<dbg> bmi270: bmi270_get_gyro_gain_update_status: Read feature reg[0x38]@0 = 0x0000
<dbg> bmi270: bmi270_get_gyro_gain_update_status: Status in x-axis: 0 y-axis: 0 z-axis 0, gtrigger: 0
<inf> bmi270: CRT was successful!
<inf> bmi270: Gyroscope user gain correction, X: 3 Y: 114 Z: 125
<dbg> bmi270: bmi270_set_aps: advance power save mode set to: 1
<dbg> bmi270: bmi270_set_aps: advance power save mode set to: 0
<dbg> bmi270: bmi270_set_aps: advance power save mode already in the intended state
<dbg> bmi270: bmi270_feature_reg_rmw: Read feature reg[0x34]@1 = 0x0088
<dbg> bmi270: bmi270_feature_reg_rmw: Wrote feature reg[0x34]@1 = 0x0488
<inf> bmi270: Programming NVM ...
<inf> bmi270: NVM prog Completed!
<dbg> bmi270: write_config_file: writing config file base
<dbg> bmi270: bmi270_set_gyro_gain: Gyro usr gain compensation status:1
```
- Power cycling the bmi270 device and verifying the data was successfully saved, and detect the new calibration was the same:
```
*** Booting Zephyr OS build zephyr-v3.4.0-4217-g341590545ac8 ***
Device 0x93e4 name is bmi270@68
<dbg> bmi270: write_config_file: writing config file base
<inf> bmi270: Gyroscope user gain correction, X: 3 Y: 114 Z: 125
<dbg> bmi270: bmi270_set_aps: advance power save mode set to: 0
<dbg> bmi270: bmi270_set_aps: advance power save mode already in the intended state
<dbg> bmi270: bmi270_feature_reg_rmw: Read feature reg[0x32]@1 = 0x0000
<dbg> bmi270: bmi270_feature_reg_rmw: Wrote feature reg[0x32]@1 = 0x0000
<wrn> bmi270: Ensure that the device is at rest during CRT execution!
<dbg> bmi270: bmi270_set_aps: advance power save mode already in the intended state
<dbg> bmi270: bmi270_feature_reg_rmw: Read feature reg[0x32]@1 = 0x0000
<dbg> bmi270: bmi270_feature_reg_rmw: Wrote feature reg[0x32]@1 = 0x0100
<inf> bmi270: CRT running.../n
<dbg> bmi270: bmi270_set_aps: advance power save mode already in the intended state
<dbg> bmi270: bmi270_get_gyro_gain_update_status: Read feature reg[0x38]@0 = 0x0000
<dbg> bmi270: bmi270_get_gyro_gain_update_status: Status in x-axis: 0 y-axis: 0 z-axis 0, gtrigger: 0
<inf> bmi270: CRT was successful!
<inf> bmi270: Gyroscope user gain correction, X: 3 Y: 114 Z: 125
<wrn> bmi270: CRT new user-gyro gains remained the same
<dbg> bmi270: bmi270_set_aps: advance power save mode set to: 1
```

peter-axe committed:
drivers: sensors: bmi270: Add macros and prototypes for CRT support.

Fixes: #62514
This commit adds the necessary macros and function prototypes to the bmi270
driver header in order to add support for the CRT,
Component Retrimming calibration feature.
It also extends the public API for bmi270,
exposing the sensors attributes for the CRT feature.

Signed-off-by: Pedro Machado <peteraxe94@gmail.com>

peter-axe committed:
drivers: sensors: bmi270: add bmi270_soft_reset method public

Fixes: #62514
The bmi270 CRT feature requires the soft_reset method.
The soft reset must be performed by the CRT module,
after nvm programming. The function, was previously
embedded in the init function. This commit moves the
soft reset logic to a new public method:bmi270_soft_reset
and the init function now calls that method.

Signed-off-by: Pedro Machado <peteraxe94@gmail.com>

peter-axe committed:
drivers: sensors: bmi270: Add CRT feature module

Fixes: #62514
Adds support for the bmi270 feature:
Creates a specific kconfig for the feature.
Adds a method to run the crt process.
Adds a method to program the bmi270 nvm.
Adds a method to enable/disable gain compensation.

Signed-off-by: Pedro Machado <peteraxe94@gmail.com>

peter-axe committed:
drivers: sensors: bmi270: Add CRT support to sensor_attr_set

Fixes: #62514
This change allows the bmi270 CRT feature methods
to be called through Zephyr's sensor api.
Using the sensor_attr_set function, with the following args:
-SENSOR_ATTR_CALIBRATION-run the crt.
-BMI270_SENSOR_ATTR_NVM_PROG-save the CRT results in nvm.
-BMI270_SENSOR_ATTR_GAIN_COMP-enable/disable gain compensation.

Signed-off-by: Pedro Machado <peteraxe94@gmail.com>

carlescufi commented:
Thanks for the PR! Please take a look at [our commit message guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines) to find out how to format your commit messages, and at [our contribution workflow](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow) to understand how to update your Pull Request.

maxmclau commented on drivers/sensor/bmi270/bmi270.h:
```diff
@@ -105,6 +108,48 @@
+#define BMI270_GYR_CRT_CONF_RUNNING_EN     0x01
+#define BMI270_GYR_CRT_CONF_RUNNING_DIS    0x00
+/* Sensor specific attributes */
+#define BMI270_SENSOR_ATTR_NVM_PROG        (SENSOR_ATTR_PRIV_START)
```
these private attributes should be defined in an enumation within a separate file `zephyr/include/zephyr/drivers/sensor/bosch_bmi270.h`.

maxmclau replied:
hmmm, I'm not sure how those attribute defines would find their way to the application. without the separate bmi270.h file I receive an error since it is not visible to the application.

```
error: 'BMI270_SENSOR_ATTR_NVM_PROG' undeclared (first use in this function)
ret = sensor_attr_set(imu_internal_dev, SENSOR_CHAN_GYRO_XYZ, BMI270_SENSOR_ATTR_NVM_PROG,
```

SHT4x has an example of how this is implemented
https://github.com/zephyrproject-rtos/zephyr/blob/9bbf7e5a4806c158e0d5c19b816379ef02e0f3e3/include/zephyr/drivers/sensor/sht4x.h#L30

MaureenHelm replied:
Agree with @maxmclau , the sm351lt and ina3221 drivers are wrong and need to be fixed

maxmclau commented on drivers/sensor/bmi270/bmi270_crt.c:
```diff
@@ -0,0 +1,743 @@
+#define BMI270_STATUS_CHECK_POLL_PERIOD_US 10000
+
+/** @name Enum to define status of gyroscope trigger G_TRIGGER */
+typedef enum {
```
I'm not sure where Zephyr documents these things, but they don't use tyepdef declarations. Enumeration members should also be prepended with the enumeration name to prevent naming collisions

```
enum {
	BMI270_CRT_TRIGGER_STAT_SUCCESS = 0,
	...
};
```

peter-axe committed:
drivers: sensors: bmi270: Add CRT support to sensor_attr_get

Fixes: #62514
This change allows using sensor_attr_get with SENSOR_ATTR_CALIBRATION
to retrieve the compensated user-gain updated data of gyroscope.
Also, it allows checking if CRT has ever been performed.

Signed-off-by: Pedro Machado <peteraxe94@gmail.com>

MaureenHelm commented on drivers/sensor/bmi270/Kconfig:
```diff
@@ -62,4 +62,10 @@ config BMI270_THREAD_STACK_SIZE
 
+config BMI270_CRT
+	bool "CRT feature support for bmi270"
+	default n
```
Please amend the commit where it's introduced instead

MaureenHelm commented on drivers/sensor/bmi270/Kconfig:
```diff
@@ -62,4 +62,10 @@ config BMI270_THREAD_STACK_SIZE
+	bool "CRT feature support for bmi270"
+	default n
+	help
+	  Gyro CRT calibration feature support for bmi270
```
Please amend the commit where it's introduced instead

MaureenHelm commented on drivers/sensor/bmi270/bmi270_crt.c:
```diff
@@ -0,0 +1,747 @@
+	BMI270_CRT_TRIGGER_STAT_DL_ERR = 0x02,
+	/*Command is aborted by host or due to motion detection*/
+	BMI270_CRT_TRIGGER_STAT_ABORT_ERR = 0x03
+} GTriggerStatus;
```
Please amend the commit where it's introduced instead

teburd commented on include/zephyr/drivers/sensor/bmi270.h:
```diff
@@ -0,0 +1,109 @@
+#define BMI270_GYR_GAIN_DIS                0x00
+
+/* Helper Macros for setting gyro gain compensation */
+#define SENSOR_VALUE_WITH_GAIN(var_name, gain_value) \
```
These seem to be BMI270 specific helper macros without any prefix to them which might lead to some confusion

teburd commented on include/zephyr/drivers/sensor/bmi270.h:
```diff
@@ -0,0 +1,109 @@
+	BMI270_CRT_TRIGGER_STAT_DL_ERR = 0x02,
+	/*Command is aborted by host or due to motion detection*/
+	BMI270_CRT_TRIGGER_STAT_ABORT_ERR = 0x03
+} gtrigger_status;
```
Is this typedef really needed or can it be used directly as an enum?

teburd commented on include/zephyr/drivers/sensor/bmi270.h:
```diff
@@ -0,0 +1,109 @@
+ * Describes the saturation status for the gyroscope gain
+ * update and G_TRIGGER command status
+ */
+struct bmi2_gyr_user_gain_status {
```
How is this struct used from an application, I see it in the private driver bmi270.h header as a member but its unclear how this is used outside of the driver

teburd commented on include/zephyr/drivers/sensor/bmi270.h:
```diff
@@ -0,0 +1,109 @@
+};
+
+/** @name Structure to store the compensated user-gain data of gyroscope */
+struct bmi2_gyro_user_gain_data {
```
Same question as above really, its unclear to me how this might be used in an application as its being placed in a public header

MaureenHelm commented:
@peter-axe could you rebase please to re-trigger CI?

avisconti reviewed (changes requested):
In my opinion the last 3 commits should be squashed together. Don't know what @MaureenHelm's opinion is.

MaureenHelm commented:
> Hi @MaureenHelm , just rebased. Do you feel that the last 3 commits should be squashed together, like @avisconti suggested?

It's fine with me either way.

Looks like some CI checks never returned, can you rebase to kick CI please?

MaureenHelm commented:
Need to rebase and resolve conflicts

MaureenHelm reviewed (approved):
@avisconti please revisit
